File: 383-Ransom Note-赎金信/383-Ransom Note-赎金信.js

/**
 * @param {string} ransomNote
 * @param {string} magazine
 * @return {boolean}
 */
 var canConstruct = function(ransomNote, magazine) {
    if(ransomNote.length > magazine.length) return false
    let map = new Map()
    for(let i = 0; i < ransomNote.length; i++){
        if(map.has(ransomNote[i])){
            map.set(ransomNote[i], map.get(ransomNote[i]) + 1)
        }else{
            map.set(ransomNote[i], 1)
        }
    }

    for(let i = 0; i < magazine.length; i++){
        if(map.has(magazine[i])){
            let count = map.get(magazine[i])
            if(count - 1 === 0){
                map.delete(magazine[i])
            }else{
                map.set(magazine[i], count - 1)
            }
        }
    }
   return map.size === 0 ? true : false
};

File: 383-Ransom Note-赎金信/383-Ransom Note-赎金信.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        int[] letters = new int[26];
    	for(char c : ransomNote.toCharArray()) {
    		letters[c- 'a']++;
    	}
    	for(char c : magazine.toCharArray()) {
    		if(letters[c - 'a'] > 0){
                letters[c - 'a']--;
            }
    	}
    	for(int i : letters) {
    		if(i != 0) return false;
    	}
    	return true;
    }
}